src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs

```diff
@@ -236,6 +237,204 @@ public static void Main()
+            using (var peReader = new PEReader(peImage))
+            {
+                var metadataReader = peReader.GetMetadataReader();
+                foreach (var method in metadataReader.MethodDefinitions)
+                {
+                    var rva = metadataReader.GetMethodDefinition(method).RelativeVirtualAddress;
+                    if (rva != 0)
+                    {
+                        var il = peReader.GetMethodBody(rva).GetILBytes();
+                        var throwNull = new[] { (byte)ILOpCode.Ldnull, (byte)ILOpCode.Throw };
+
+                        if (expectEmptyOrThrowNull)
+                        {
+                            AssertEx.Equal(throwNull, il);
+                        }
+                        else
+                        {
+                            AssertEx.NotEqual(throwNull, il);
+                        }
+                    }
+                }
+            }
+        }
+
+        [Fact]
+        public void EmitMetadataOnly_NoPdbs()
+        {
+            CSharpCompilation comp = CreateCompilation("", references: new[] { MscorlibRef },
+                options: TestOptions.DebugDll.WithDeterministic(true));
+
+            using (var output = new MemoryStream())
+            using (var pdbOutput = new MemoryStream())
+            {
+                Assert.Throws<ArgumentException>(() => comp.Emit(output, pdbOutput,
+                    options: EmitOptions.Default.WithEmitMetadataOnly(true)));
+            }
+        }
+
+        [Fact]
+        public void EmitMetadataOnly_NoMetadataPeStream()
+        {
+            CSharpCompilation comp = CreateCompilation("", references: new[] { MscorlibRef },
+                options: TestOptions.DebugDll.WithDeterministic(true));
+
+            using (var output = new MemoryStream())
+            using (var metadataPeOutput = new MemoryStream())
+            {
+                Assert.Throws<ArgumentException>(() => comp.Emit(output, metadataPeStream: metadataPeOutput,
+                    options: EmitOptions.Default.WithEmitMetadataOnly(true)));
+            }
+        }
+
+        [Fact]
+        public void EmitMetadata()
+        {
+            string source = @"
+public abstract class PublicClass
+{
+    public void PublicMethod() { System.Console.Write(""Hello""); }
+}
+";
+            CSharpCompilation comp = CreateCompilation(source, references: new[] { MscorlibRef },
+                options: TestOptions.DebugDll.WithDeterministic(true));
+
+            using (var output = new MemoryStream())
+            using (var pdbOutput = new MemoryStream())
+            using (var metadataOutput = new MemoryStream())
+            {
+                var result = comp.Emit(output, pdbOutput, metadataPeStream: metadataOutput);
+                Assert.True(result.Success);
+                Assert.NotEqual(0, output.Position);
+                Assert.NotEqual(0, pdbOutput.Position);
+                Assert.NotEqual(0, metadataOutput.Position);
```

Comment:
Consider adding `VerifyMethodBodies(ImmutableArray.CreateRange(output.GetBuffer()), expectedEmptyOrThrowNull: true)`, so it's clear there's a difference between this case and the `EmitToArray()` case below.